src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -539,12 +549,51 @@ public long importKey(long hSession, CK_ATTRIBUTE[] attributes) throws PKCS11Exc
             if (attr.type == CKA_KEY_TYPE) {
                 keyType = attr.getLong();
             }
-            if (attr.type == CKA_VALUE) {
-                keyBytes = attr.getByteArray();
+            ckAttrsMap.put(attr.type, attr);
+        }
+
+        if(keyClass == CKO_PRIVATE_KEY) {
+            if(keyType == CKK_RSA) {
+                try {
+                    keyBytes = RSAPrivateCrtKeyImpl.newKey(
+                        KeyType.RSA,
+                        null,
+                        ckAttrsMap.getOrDefault(CKA_MODULUS, new CK_ATTRIBUTE(CKA_MODULUS, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PUBLIC_EXPONENT, new CK_ATTRIBUTE(CKA_PUBLIC_EXPONENT, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIVATE_EXPONENT, new CK_ATTRIBUTE(CKA_PRIVATE_EXPONENT, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIME_1, new CK_ATTRIBUTE(CKA_PRIME_1, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIME_2, new CK_ATTRIBUTE(CKA_PRIME_2, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_EXPONENT_1, new CK_ATTRIBUTE(CKA_EXPONENT_1, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_EXPONENT_2, new CK_ATTRIBUTE(CKA_EXPONENT_2, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_COEFFICIENT, new CK_ATTRIBUTE(CKA_COEFFICIENT, BigInteger.ZERO)).getBigInteger()
```

Comment:
Suggest creating a helper method to avoid creating `CK_ATTRIBUTE` objects needlessly (and reducing the code volume).

Reply:
updated